Enable use of shields

erwango opened the pull request:
Simplify usage of shields within Zephyr by providing an integrated way to configure them
and add them easily to a compatible board, with the minimum dependency (ie: support of matching connector).

### Shield configuration and activation
A boards/shields/ directory is created to store the shields ports. For each shield, Kconfig, dts overlay, dts fixup and documentation should be provided ("shield_defconfig" file is not supported)
At shield activation (setting matching config flag to "y"), shield config is added to project configuration and dts.overlay piece is merged with board dts.

### Shield support
In order to support a shield, a board should be pin compatible and provide connector support in its dts file. With upcoming dtc v1.4.2, this will be made possible thanks to newly introduced overriding node element, which allows to bind generic adapter with board specific nodes. For instance, in the case of a board supporting Arduino connector and for which Arduino I2C is I2C1, it could be specified as follows in board dts file:
`arduino_i2c: i2c1 {};`

Since dtc v1.4.2 is not yet supported in Zephyr SDK, a temporary alternate way to enable connector in board dts file is available:
`#define arduino_i2c i2c1`
`	aliases {
		arduino,i2c = &i2c1;
	};`

Board nucleo_f429zi has been modified to enable i2c_arduino, as proposed above.

### Ported shields
Two similar shields are ported in in this PR order to validate possible usage of multiple (compatible) shields:
-x-nucleo-iks01a1
-x-nucleo-iks01a2
A sample is provided to use x-nucleo-iks01a1.

## PR Features:
- [x] Enable use of over-rinding nodes [1] for upcoming dtc V1.4.2
- [x] Use of multiple shields possible
- [x] x-nucleo-iks01a1 and x-nucleo-iks01a2 porting (including doc)
- [x] Provide x-nucleo-iks01a1 sample
- [x] Update Zephy documentation
- [x] nucleo_f429zi arduino I2C suport

[1] https://git.kernel.org/pub/scm/utils/dtc/dtc.git/commit/?h=v1.4.6&id=3346e065aacb1f05bdfb3872e3ba709b5f895846

erwango committed:
boards: add support for shields x-nucleo-iks01a1 and -iks01a2

Provide definition files for x-nucleo-iks01a1 and
x-nucleo-iks01a2 arduino shields

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
build: enable shields dir treatment

In dts.cmake, build shield list. For each shield, check if
associate KCONFIG is enabled and if it is, add matching files
to following new cmake variables:
-DTC_SHIELD_OVERLAY_FILE
-DTC_SHIELD_FIXUP
Then, consume these variables when needed.

Shield overlay is conceived as the very first overlay applied
to the board dts. It intends to build a new board that should
then behave as any other board versus common or user's overlay.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

SebastianBoe commented:
A minor comment, is this support specific to Arduino in any way?

If not, what about calling this "daughterboards" rather than "shields", as "shield" is a Arduino term for an Arduino daughterboard.

erwango commented:
> If not, what about calling this "daughterboards" rather than "shields", as "shield" is a Arduino term for an Arduino daughterboard.

Indeed this should not be arduino specific. Though, I find daughterboard a bit long and heavy to use (quite a subjective point). Any alternative? Expansion? mezzanine (if not seen as 96board specific )?

erwango commented:
Pushed a new version enabling  treatment of dtc 'overriding nodes'.
They are detected and generated by devicetree.py as 'alt_name' of a node and then treated as aliases in dts_extract_includes.py.
Then I made some changes in dts_extract_includes.py to generate bindings #defines using aliases to get generic references.
This require dtc V1.4.2 (I hacked dts.cmake to use local dtc version)

mike-scott commented on boards/shields/x_nucleo_iks01a1/Kconfig.shield:
```diff
@@ -0,0 +1,15 @@
+# Kconfig - 96boards Carbon STMF401RE board configuration
```
Looks like a copy / paste error here referring to 96boards Carbon

mike-scott reviewed (commented):
Minor comment issue

mike-scott commented:
I'm interested in this PR as I'm working with an Arduino header compatible LTE-M modem.

Adding to the discussion of current terms:
**Daughterboard**: quite long
**Daughtercard**: same
**Shield**: Arduino-specific term for daughterboard.  Agree that this could be confusing as we would be using the term for broader use.
**Mezzanine**: I don't think this is a 96 boards specific term, however, it does refer to a board that is stacked on top of and parallel to the main board.  Still a good choice in my opinion.

Adding this as an option:
hat (Hardware Attached on Top): Raspberry Pi-related term.  Could cause the same confusion as shield.  Not sure if any of the other prototyping boards have picked this term up.

erwango commented:
I had a look to google and it appears "mezzanine" is rather 96boards tainted, even if also used as generic.

What about "expansion"?

nashif commented:
imo shield is just fine.

nashif commented:
sorry, closed by mistake

SebastianBoe commented:
"Is there a way to amend .config file with <shield>_defconfig file once it is known?"

No.

It sounds like you are saying that devicetree is trying to give input to Kconfig. This is a problem because in the current design devicetree is done after Kconfig.

Intuitively, one would expect that it would be more natural to do HW-description first, and then SW-configuration based on an unconfigurable HW-description.

So one would actually expect that devicetree was processed before Kconfig. Re-ordering Kconfig and devicetree might solve your problem, but I don't know what was behind the design decision to put devicetree after Kconfig, so I don't know what kind of consequences that will have.

erwango commented:
> It sounds like you are saying that devicetree is trying to give input to Kconfig

Not exactly. Since we want to be able to stack multiple shields, we're using Kconfig to know which shield is enabled(CONFIG_SHIELD_XXX = y) to input additional files (dts overlay, dts fixup and additional CONFIG_ bits) into build.

> Re-ordering Kconfig and devicetree might solve your problem

It won't help neither because we need Kconfig to decide which .overlay should be applied to dts.

Setting aside dts for a while, is there a way to parse _defconfig files depending on 
other CONFIG_ flags

SebastianBoe commented:
"Setting aside dts for a while, is there a way to parse defconfig files depending on
other CONFIG flags"

If I understand the question correctly, it is possible to do:

```
if(CONFIG_IMPORT_ANOTHER_KCONFIG_FRAGMENT_INTO_THE_CMAKE_NAMESPACE)
  # Parse the lines prefixed with CONFIG_ in the .config file from Kconfig
  import_kconfig(${CONFIG_EXTRA_KCONFIG_FRAGMENT})
endif()

print(CONFIG_IMPORTED_VALUE)

```

But I cannot imagine this is desirable. Because Kconfig will no longer be checking the validity of the configuration. So as far as I can understand the question, no, this is not possible.

erwango commented:
> If I understand the question correctly

It seems yes.
I see another way to do it, not sure it is acceptable though:
In Kconfig.cmake, after import_kconfig(${DOT_CONFIG})
do the same processing as proposed in dts.cmake (parse shields/ , check CONFIG_SHIELD_XX flags,
if defined, populate a  ${SHIELD_DEFCONFIG}  ,  go back to start of .config generation (set merge_config_files...), and try not to loop forever..

SebastianBoe commented:
We are only doing 1 Kconfig pass per build today and I fear introducing a multi-pass Kconfig would create unwanted complexity.

erwango commented:
> We are only doing 1 Kconfig pass per build today and I fear introducing a multi-pass Kconfig would create unwanted complexity.

Understood.
Ok. Then it seems we'll have one limitation, which is unability to provide a shield _deconfig file.
This is unfortunate, but I guess we can live with it.
For most of flags this is not an issue, as they can be provided through Kconfig.shield or Kconfig.defconfig.
Only issue is for CONFIG_ flags generated from a "choice" kconfig directive, as they don't work well in  Kconfig.defconfig files. But maybe this will get better one day.

b0661 commented:
> scripts/extract/globals: treat node alternate names as 'aliases'

> Treat new generated dtc element 'alt-name' as usual aliases.
> ..
> For instance, defining following overriding node:
> arduino_i2c : i2c1 {};
> Will generate following definition:
> #define ARDUINO_I2C_LABEL	ST_STM32_I2C_V1_40005400_LABEL

You may just generate

  #define ARDUINO_I2C ST_STM32_I2C_V1_40005400

as adding the property part (e.g. LABEL) can easily be done by the driver. That way it is also clear it is an alternate name.

erwango commented:
> You may just generate

> #define ARDUINO_I2C ST_STM32_I2C_V1_40005400

> as adding the property part (e.g. LABEL) can easily be done by the driver. That way it is also clear it is an alternate name.

Not sure I get your point. Actually my goal is to leave the driver untouched so both set of definitions could co-exist and be used at same time by different applications.
Anyway, I'll push this alias generation part as a separate PR to ease discussions.

b0661 commented:
> Not sure I get your point. Actually my goal is to leave the driver untouched so both set of definitions could co-exist and be used at same time by different applications

A user may yust do CONCAT(ARDUINO_I2C, _LABEL) to access the label.

erwango commented:
Status:
When using dtc v1.4.2, use of shields is now fully functional.
Only limitation is that no _defconfig file could be used, but with #6966, all types of CONFIG_ flags could be set in Kconfig.defconfig files, which makes _defconfig useless (at least in shield specific case).

Remaining work is to document properly and provide a dedicated sample application.

galak commented:
> imo shield is just fine.

I think shield is fine as well.

erwango committed:
samples: Provide x-nucleo-iks01a1 shield sample

This sample is made to demonstrate use of shield x-nucleo-ik01a1.
It will display embedded sensor data endlessly.
It requires a board with Arduino i2c as minimum configuration

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
doc: provide shield usage documentation

Following introduction of shield feature,
provide requested documentation.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented:
PR is now complete and ready for broad review.

In order to enable use today without dtc v1.4.2 support in Zephyr SDK, an alternate method for
device tree overriding node (which allows merge of board+shield dts) as been provided.

dbkinder commented on doc/porting/shields.rst:
```diff
@@ -0,0 +1,83 @@
+Shields
+#######
+
+Shields have been introduced on Zephyr as a new element that can be used
```
This won't be a "new element" in a few months, so better to say:

    Shields, also known as "add-on" or "daughter boards", attach to a board
    to extend its features and services for easier and modularized prototyping.
    In Zephyr, the shield feature provides Zephyr-formatted shield
    descriptions for easier compatibility with applications.

dbkinder commented on doc/porting/shields.rst:
```diff
@@ -0,0 +1,83 @@
+Shield porting and configuration
+********************************
+
+Shields are available in board directory under :file:`/boards/shields`.
```
I'd merge these two sentences and say:

    Shield configuration files are available in the board directory 
    under :file:`/boards/shields`:

dbkinder commented on doc/porting/shields.rst:
```diff
@@ -0,0 +1,83 @@
+These files provides shield configuration as follows:
+
+* **Kconfig.shield**: This file defines the shield's config flag
+  (CONFIG_SHIELD_X). This flag will be used in an application to trigger the
```
Better to talk in the present rather than the future:

    Applications use this flag to trigger the shield configuration
    on top of the board configuration, enabling each shield
    component or sensor driver.

dbkinder commented on doc/porting/shields.rst:
```diff
@@ -0,0 +1,83 @@
+  shield configuration on top of board configuration. Besides in this file,
+  each driver of shield components (sensors for instance) are enabled.
+
+* **Kconfig.defconfig**: This file is used to provide specific driver default
```
change to 

    This file provides conditional configuration of specific driver
    default configurations, offering flexibility for broad board support.

dbkinder commented on doc/porting/shields.rst:
```diff
@@ -0,0 +1,83 @@
+  configuration. Since congiruration could be done conditionally it provides
+  flexibility for broader board support.
+
+* **<shield>.overlay**: Device tree overlay is a device tree fragment that
```
change to:

    This file provides a shield description in device tree format that is
    merged with the board's device tree information before compilation.

dbkinder commented on doc/porting/shields.rst:
```diff
@@ -0,0 +1,83 @@
+  provides shield description in device tree format. This file will be merged
+  with board device tree file before compilation.
+
+* **dts.fixup**: fixup file to bind board components definitions with
```
Add  ``This is a fixup file ...`` (to be consistent with the previous items)

dbkinder commented on doc/porting/shields.rst:
```diff
@@ -0,0 +1,83 @@
+Board compatibility
+*******************
+
+Shields to board compatibility relies on use of well-known connectors populated
```
Hardware shield-to-board compatibility depends on the use of well-known
    connectors used on popular boards (such as Arduino and 96boards).  For
    software compatibility, boards must also provide a configuration matching
    their supported connectors.

dbkinder commented on doc/porting/shields.rst:
```diff
@@ -0,0 +1,83 @@
+
+This should be done at two different level:
+
+* Pinmux: Connector pins should be correctly configured to match shield pining
```
change ``pining`` to ``pins``

dbkinder commented on doc/porting/shields.rst:
```diff
@@ -0,0 +1,83 @@
+
+* Pinmux: Connector pins should be correctly configured to match shield pining
+
+* Device tree: Board device tree file should propose a node alias for each
```
A board device tree file should define a node alias for each
    connector interface.  For example, for Arduino I2C:

dbkinder commented on doc/porting/shields.rst:
```diff
@@ -0,0 +1,83 @@
+                arduino,i2c = &i2c1;
+        };
+
+Note: With support of dtc v1.4.2, above will be raplced with recently introduced
```
``replaced with the recently``

dbkinder commented on doc/porting/shields.rst:
```diff
@@ -0,0 +1,83 @@
+Shield activation
+*****************
+
+To activate the support of a shield with a board, once will need to define
```
Activate support for a shield by enabling the corresponding Kconfig shield option:

dbkinder commented on doc/porting/shields.rst:
```diff
@@ -0,0 +1,83 @@
+
+.. code-block:: none
+
+        #CONFIG_<shield>=y
```
Maybe provide a specific example here?

dbkinder commented on samples/shields/x_nucleo_iks01a1/README.rst:
```diff
@@ -0,0 +1,57 @@
+.. _x-nucleo-iks01a1-sample:
+
+X-NUCLEO-IKS01A1: MEMS Interial and Environmental Multi sensor shield
```
X-NUCLEO-IKS01A1: MEMS inertial and environmental multi-sensor shield

dbkinder commented on samples/shields/x_nucleo_iks01a1/README.rst:
```diff
@@ -0,0 +1,57 @@
+
+Overview
+********
+This sample enables all sensors of X-NUCLEO-IKS01A1 shield by reading
```
This sample enables all sensors of a X-NUCLEO-IKS01A1 shield, and then
    periodically reads and displays data from the shield sensors:

    - HTS221: Temperature and humidity
    - LPS25HB: Atmospheric pressure
    - LIS3MDL: 3-axis Magnetic field intensity
    - LSM6DSL: 3-Axis Acceleration

dbkinder commented on samples/shields/x_nucleo_iks01a1/README.rst:
```diff
@@ -0,0 +1,57 @@
+Requirements
+************
+
+This sample uses the X-NUCLEO-IKS01A1 shield stacked on a board over
```
This sample communicates over I2C with the X-NUCLEO-IKS01A1 shield
    stacked on a board with an Arduino connector. The board's I2C must be
    configured for the I2C Arduino connector (both for pin muxing
    and device tree).

dbkinder commented on samples/shields/x_nucleo_iks01a1/README.rst:
```diff
@@ -0,0 +1,57 @@
+References
+**********
+
+ - X-NUCLEO-IKS01A1: http://www.st.com/en/ecosystems/x-nucleo-iks01a1.html
```
delete the leading space before the ``-``

dbkinder commented on samples/shields/x_nucleo_iks01a1/README.rst:
```diff
@@ -0,0 +1,57 @@
+Building and Running
+********************
+
+This sample runs with X-NUCLEO-IKS01A1 stacked on any board with matching
```
``with a matching``

dbkinder commented on samples/shields/x_nucleo_iks01a1/README.rst:
```diff
@@ -0,0 +1,57 @@
+********************
+
+This sample runs with X-NUCLEO-IKS01A1 stacked on any board with matching
+Arduino connector. For this example, board used is nucleo_f429zi.
```
For this example, we use a :ref:`nucleo_f429zi_board` board.

dbkinder commented on samples/shields/x_nucleo_iks01a1/README.rst:
```diff
@@ -0,0 +1,57 @@
+.. _x-nucleo-iks01a1-sample:
```
You'll also need to add a line in ``samples/samples.rst`` to include ``shields/*`` to get his new sample tree included in the generated docs.

erwango replied:
Indeed, had missed that, thanks!

dbkinder commented on boards/shields/x_nucleo_iks01a1/doc/x_nucleo_iks01a1.rst:
```diff
@@ -0,0 +1,57 @@
+***********
+
+An example on how to use the ``x-nucleo-iks01a1`` shield is available
+with :ref:`x-nucleo-iks01a1-sample` application (see :ref:`shield_use` for
```
the label ``shield_use`` isn't defined anywhere?

dbkinder reviewed (changes requested):
Thanks for getting the docs started...

erwango commented:
Thanks @dbkinder for the review, I think I addressed all your comments

erwango commented:
@avisconti : Might be of some interest for you

dbkinder commented on boards/shields/x_nucleo_iks01a1/doc/x_nucleo_iks01a1.rst:
```diff
@@ -0,0 +1,63 @@
+humidity and temperature sensor, and the LPS25HB pressure sensor.
+
+The X-NUCLEO-IKS01A1 interfaces with the STM32 microcontroller via the I2C pin,
+and it is possible to change the default I3C port.
```
``I2C``

dbkinder commented on doc/porting/shields.rst:
```diff
@@ -0,0 +1,82 @@
+*****************
+
+Activate support for a shield by enabling the corresponding Kconfig shield
+option:
```
for clarity, which dts file should this be added to?

dbkinder commented on samples/shields/x_nucleo_iks01a1/README.rst:
```diff
@@ -0,0 +1,55 @@
+References
+**********
+
+ -X-NUCLEO-IKS01A1: http://www.st.com/en/ecosystems/x-nucleo-iks01a1.html
```
delete the leading space before the ``-``

dbkinder reviewed (changes requested):
Getting close now.  Thanks.

SebastianBoe commented on boards/arm/nucleo_f429zi/nucleo_f429zi.dts:
```diff
@@ -7,6 +7,9 @@
 /dts-v1/;
 #include <st/stm32f429.dtsi>
 
+/* Temp: define arduino_i2c */
```
There should exist a github issue for lack of the DTC feature support in the SDK and this comment should be stating that this is a workaround around that Github issue.

SebastianBoe commented on boards/shields/x_nucleo_iks01a1/doc/x_nucleo_iks01a1.rst:
```diff
@@ -0,0 +1,63 @@
+humidity and temperature sensor, and the LPS25HB pressure sensor.
+
+The X-NUCLEO-IKS01A1 interfaces with the STM32 microcontroller via the I2C pin,
+and it is possible to change the default I2C port.
```
When I read this it sounds weird.

Perhaps it should say:

The X-NUCLE0-IKS01A1 board allows other boards to interface with the STM32 via an I2C interface.

dbkinder replied:
I like!  (The current line is a paste http://www.st.com/en/ecosystems/x-nucleo-iks01a1.html)

SebastianBoe commented on cmake/dts.cmake:
```diff
@@ -13,10 +13,48 @@ set(GENERATED_DTS_BOARD_CONF ${PROJECT_BINARY_DIR}/include/generated/generated_d
+
+#Parse boards/shields to generate the shield list
+
+set(shield_dir $ENV{ZEPHYR_BASE}/boards/shields)
```
This must be ${ZEPHYR_BASE} instead.

erwango replied:
ok, thanks

SebastianBoe commented on cmake/dts.cmake:
```diff
@@ -13,10 +13,48 @@ set(GENERATED_DTS_BOARD_CONF ${PROJECT_BINARY_DIR}/include/generated/generated_d
+
+  # Generate CONFIG flags matching each shield
+  string(TOUPPER config_shield_${shield} UPPER_CASE_CONFIG_SHIELD)
+  set(shield_config ${UPPER_CASE_CONFIG_SHIELD})
```
It is is cleaner to do this oneliner instead of those two statements:

string(TOUPPER "CONFIG_SHIELD_${shield}" shield_config)

erwango replied:
Done, thanks

SebastianBoe commented on cmake/dts.cmake:
```diff
@@ -28,6 +66,7 @@ if(CONFIG_HAS_DTS)
   set(
     dts_files
     ${DTS_SOURCE}
+    ${DTC_SHIELD_OVERLAY_FILES_AS_LIST}
```
We can't modify core build scripts every time there is a new use-case. We must strive to re-use a few existing concepts instead of adding more and more complexity.

Could you look into if it is possible, to get the shield overlay files into DTC_OVERLAY_FILE_AS_LIST somehow to avoid this (and similair) patches?

And if not possible, elaborate on why not it cannot be done so we better understand the limitations of the DTS infrastructure. (And so that reviewers can easily verify that we have done all that is possible to minimize complexity).

erwango replied:
Agreed. I added this new variable initially for a better control of my change, and be sure I don't mess with DTC_OVERLAY_FILE variable precedence. But now I should be able to do it w/o DTC_OVERLAY_FILE_AS_LIST

dbkinder reviewed (approved):
+1 (but I also like @SebastianBoe 's suggestion)

SebastianBoe committed:
dts: Simplify overlay build scripts

Miscellaneous refactorings to simplify the overlay build scripts.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

SebastianBoe commented:
@erwango : Thanks. I had some minor comments to the new version, decided to put the feedback in a PR: https://github.com/erwango/zephyr/pull/2

SebastianBoe commented:
Approved assuming https://github.com/erwango/zephyr/pull/2 is merged

erwango commented:
@SebastianBoe : Thanks. I added your commit in the branch

erwango commented:
@mike-scott , as potential user, are you ok with current PR?

mike-scott commented:
@erwango I have a working setup using your patches to support the AT&T IoT Starter Kit (LTE-M) cellular shield for K64F.

I realize that for my purposes it's only a marginal improvement.  Other than the UART there aren't any other peripherals exposed by the shield.

With your patches:
1. I'm able to abstract the UART device label entirely to dts + fixup.dts.
2. Move the board specific "select xxx" Kconfig settings to boards/shields/avt_wcn14a2a_v1/Kconfig.shield instead of the driver/modems/Kconfig which consolidates all of the driver CONFIG settings under CONFIG_SHIELD_AVT_WCN14A2A_V1

It seems like a good start, that can be improved upon once a solution for DTS-based pinmux / GPIOs becomes available.

NOTE: It will still be tricky as Arduino pin mappings will need to be established *per pin* across several GPIO ports, but that's a problem for a different day.

erwango commented:
@galak , as discussed, added connector compatibility flags.

One minor issue is Error message in case of incompatibility (flag not selected by board):
```
-- Generating zephyr/include/generated/generated_dts_board.h
Error: nucleo_f429zi.dts.pre.tmp:496.1-6 syntax error
FATAL ERROR: Unable to parse input tree
CMake Error at /local/mcu/zephyr/zephyr-project/cmake/dts.cmake:118 (message):
  command failed with return code: 1
Call Stack (most recent call first):
  /local/mcu/zephyr/zephyr-project/cmake/app/boilerplate.cmake:263 (include)
  CMakeLists.txt:1 (include)
```
This is not quite explicit, but I didn't find a way to get it better. I take any idea to get it more explicit.

erwango commented:
@mike-scott , thanks for the feedback, maybe #7151 will be to put this a step further.

mike-scott commented:
@erwango would be great to see what this PR looks like now with some of it's commits merged into master

erwango commented:
@mike-scott, rebased. Sadly I missed 1.13 merge window.
Hope we'll get it merged in early 1.14

mike-scott commented:
@erwango no worries.  I will continue to test this as needed. :)

erwango commented:
@galak , can we merge this one now?

erwango commented:
@galak: rebased vs recent soc changes and removed connector flags as it was introducing CONFIG_ in dts which we're now wanting to avoid.
In case a bord doesn't support matching connector, dtc triggers an error:
```
-- Generating zephyr/include/generated/generated_dts_board.h
Error: nucleo_f401re.dts.pre.tmp:459.1-13 Label or path arduino_i2c not found
FATAL ERROR: Syntax error parsing input tree
```
I think this should be explicit enough.

galak commented on boards/arm/nucleo_f429zi/nucleo_f429zi.dts:
```diff
@@ -47,6 +47,9 @@
 		led2 = &red_led_1;
 		sw0 = &user_button;
+		/* Temp provide arduino i2c alias */
+		/* Workaround waiting meta-zephyr-sdk/issues/92 resolution */
```
Do we still need this, since the SDK now supports a good version of DTC?

erwango replied:
Good catch! Not needed anymore indeed.
Fixed, thanks!

galak reviewed (dismissed):
Minor comment.

MaureenHelm commented on samples/shields/x_nucleo_iks01a1/src/main.c:
```diff
@@ -0,0 +1,103 @@
+	struct device *hts221 = device_get_binding("HTS221");
+	struct device *lis3mdl = device_get_binding("LIS3MDL");
+	struct device *lsm6ds0 = device_get_binding("LSM6DS0");
+	struct device *lps25hb = device_get_binding("LPS25HB");
```
Get these from the device tree labels rather than hardcoding the strings

erwango replied:
Indeed, would better, thanks for the review and hint! Done!

MaureenHelm reviewed (commented):
Just one minor nit, but otherwise this looks really nice.
